ext/readline/readline_cli.c

```diff
@@ -349,9 +349,14 @@ static int cli_is_valid_code(char *code, size_t len, zend_string **prompt) /* {{
 			case heredoc:
 				if (code[i - (heredoc_len + 1)] == '\n' && !strncmp(code + i - heredoc_len, heredoc_tag, heredoc_len) && code[i] == '\n') {
 					code_type = body;
-				} else if (code[i - (heredoc_len + 2)] == '\n' && !strncmp(code + i - heredoc_len - 1, heredoc_tag, heredoc_len) && code[i-1] == ';' && code[i] == '\n') {
+				} else if (!strncmp(code + i - heredoc_len + 1, heredoc_tag, heredoc_len)) {
+					char c = code[i + 1];
```

Comment:
Do we need to check string length here?

Reply:
If we get there, a `strncmp()` starting at `code[i+1-heredoc_len]` has matched for a size of `heredoc_len`, so `code[i+1]` should be always valid.

Reply:
The strncmp only inspects memory up to code+i, so I don't think that really tells us anything. There seems to be a bunch of other code using i+1 lookahead though, so I'll assume that these strings are guaranteed null terminated and thus safe to access one past end.

Reply:
You're right, the `strcmp` doesn't access `code[i+1]`, but there should always be a trailing line break (seems to always be LF even on Windows), so I don't see a problem here.